src/xdg-autostart-generator/xdg-autostart-service.c

```diff
@@ -0,0 +1,581 @@
+                "SourcePath=%s\n"
+                "PartOf=graphical-session.target\n\n",
+                path_escaped);
+
+        if (service->description) {
+                _cleanup_free_ char *t = NULL;
+
+                t = specifier_escape(service->description);
+                if (!t)
+                        return log_oom();
+
+                fprintf(f, "Description=%s\n", t);
+        }
+
+        /* Only start after the session is ready.
+         * XXX: GNOME has an autostart order which we may want to support.
+         *      It is not clear how this can be implemented reliably, which
+         *      is why it is skipped for now. */
+        fprintf(f,
+                "After=graphical-session.target\n");
+
+        fprintf(f,
+                "\n[Service]\n"
+                "Type=simple\n"
+                "ExecStart=:%s\n"
+                "Restart=no\n"
+                "TimeoutSec=5s\n"
+                "Slice=app.slice\n",
+                exec_start);
+
+        /* Generate an ExecCondition to check $XDG_CURRENT_DESKTOP */
+        if (!strv_isempty(service->only_show_in) || !strv_isempty(service->not_show_in)) {
+                _cleanup_free_ char *only_show_in = NULL, *not_show_in = NULL, *e_only_show_in = NULL, *e_not_show_in = NULL;
+
+                only_show_in = strv_join(service->only_show_in, ":");
+                not_show_in = strv_join(service->not_show_in, ":");
+                if (!only_show_in || !not_show_in)
+                        return log_oom();
+
+                e_only_show_in = cescape(only_show_in);
+                e_not_show_in = cescape(not_show_in);
+                if (!e_only_show_in || !e_not_show_in)
+                        return log_oom();
+
+                /* Just assume the values are reasonably sane */
+                fprintf(f,
+                        "ExecCondition=" ROOTLIBEXECDIR "/systemd-xdg-autostart-condition \"%s\" \"%s\"\n",
+                        e_only_show_in,
+                        e_not_show_in);
+        }
+
+        /* Generate an ExecCondition for GNOME autostart condition */
+        if (!isempty(service->autostart_condition)) {
+                _cleanup_free_ char *gnome_autostart_condition_path = NULL;
+                _cleanup_free_ char *e_autostart_condition = NULL;
+
+                if (find_binary("gnome-systemd-autostart-condition", &gnome_autostart_condition_path) < 0) {
+                        e_autostart_condition = cescape(service->autostart_condition);
+                        if (!e_autostart_condition)
+                                return log_oom();
+
+                         fprintf(f,
+                                 "ExecCondition=%s --condition \"%s\"\n",
+                                 gnome_autostart_condition_path,
+                                 e_autostart_condition);
+                 } else {
+                         log_info_errno(r, "gnome-systemd-autostart-condition not found: %m");
```

Comment:
should be log_warning_errno() i guess.

Generally:

LOG_ERROR: when we encounter an error and then abort whatever we are doing because of it
LOG_WARNING: when we encounter some issue and then continue anyway, but want to draw attention
LOG_NOTICE: when something is normal, but still of particular relevance
LOG_INFO: when something is normal and boring